Add docs for EC2 instance discovery (#16279)
gravitational/teleport: docs/pages/server-access/guides/ec2-discovery.mdx

---
title: Configure Teleport to Automatically Discover EC2 instances

> description: How to configure Teleport to install itself EC2 instances.

Reviewer: ```suggestion
description: How to configure Teleport to install itself on EC2 instances.
```

---

> Teleport's SSH Service can be configured to automatically discover, and install

Reviewer: ```suggestion
Teleport can be configured to automatically discover, and install
```

It's technically not the SSH service, right?

> Teleport on newly provisioned EC2 nodes.

Reviewer: Can we add one or two more sentences here the describes how it works? Something like:

A Teleport Process can be configured to run a discovery service, which connects to EC2, discovers instances, and runs an install script on these instances using SSM"

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- AWS account with EC2 instances and permissions to create and attach IAM policies as well as permission to create System Manager documents.

> - A running Teleport Node. See the [Server Access Getting Started Guide](../getting-started.mdx) for how to add a Node to your Teleport cluster.

Reviewer: "Node" typically means "process running the SSH service"

Your new discovery service can run anywhere right? It could run in the Teleport Auth service, for example.

## Step 1/4. Create an EC2 invite token

> When discovering EC2 nodes Teleport makes use of EC2 invite tokens for

Reviewer: I think EC2 join and IAM join are slightly different things. In the description here you say EC2, but in the YAML below you are creating `join_method: iam`.

spec:
  # use the minimal set of roles required
  roles: [Node]

  # set the join method allowed for this token
  join_method: iam

  allow:
  # specify the AWS account which nodes may join from
  - aws_account: "123456789"
```

Add the token to the Teleport cluster with:
```
tctl create -f token.yaml
```

More documentation on EC2 invite tokens can be found [here](../../management/guides/joining-nodes-aws-iam.mdx)

## Step 2/4. Create an AWS Systems Manager Document

Teleport makes use of AWS Systems Manager documents to execute commands necessary to
install and start Teleport on discovered EC2 instances.

A document with the following content should be created, named `TeleportDiscoveryInstaller`:

```yaml
---
schemaVersion: '2.2'
description: aws:runShellScript
parameters:
  token:
    type: String
    description: "(Required) The Teleport invite token to use when joining the cluster."
  scriptName:
    type: String
    description: "(Required) The Teleport instalelr script to use when joining the cluster."
mainSteps:
- action: aws:downloadContent
  name: downloadContent
  inputs:
    sourceType: "HTTP"
    destinationPath: "/tmp/installTeleport.sh"
    sourceInfo:

>       url: "https://teleport.example.com/webapi/scripts/{{ scriptName }}"

Reviewer: Are you using Go templating for this or just doing a substitute?

The syntax looks like Go templating, but if it is in fact Go templates then it should probably be `{{ .ScriptName }}`.

- action: aws:runShellScript
  name: runShellScript
  inputs:
    timeoutSeconds: '300'
    runCommand:

>       - /bin/sh /tmp/installTeleport.sh "{{ token }}"

Reviewer: There's no way for a user to provide this token, right? Teleport always provides it?

If this can be user controlled, we need to be sure to validate it, otherwise someone could provide a token of `&& rm -rf /`.

```

Here, `teleport.example.com` should be substituted with the public proxy URL of the
Teleport cluster.

When Teleport executes the System Manager document it will provide the `token`
and `scriptName` parameters with defaults.

- `{{ scriptName }}` will by default evaluate to `default-installer`
- `{{ token }}` will by default evauluate to `aws-discovery-iam-token`

## Step 3/4. Set up AWS IAM credentials

> The Teleport Discovery Node requires an IAM policy with permissions to

Reviewer: 'Discovery Node' is not a term we've ever used before. Do you think readers will understand what this means?

    sourceInfo:
      url: "https://teleport.example.com/webapi/scripts/{{ scriptName }}"
- action: aws:runShellScript
  name: runShellScript
  inputs:
    timeoutSeconds: '300'
    runCommand:
      - /bin/sh /tmp/installTeleport.sh "{{ token }}"
```

Here, `teleport.example.com` should be substituted with the public proxy URL of the
Teleport cluster.

When Teleport executes the Systems Manager document it will provide the `token`
and `scriptName` parameters with defaults.

- `{{ scriptName }}` will by default evaluate to `default-installer`
- `{{ token }}` will by default evauluate to `aws-discovery-iam-token`

## Step 3/4. Set up AWS IAM credentials

In order for Teleport to send commands to EC2 instances, the node
performing discovery requires an IAM policy with permissions to
execute Systems Manager commands and retrieve EC2 instances:

```js
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "ssm:SendCommand",
                "ec2:DescribeInstances",
                "ssm:GetCommandInvocation"
            ],
            "Resource": "*"
        }
    ]
}
```

All EC2 instances that are to be added to the Teleport cluster by the

> Discovery Node must include the `AmazonSSMManagedInstanceCore` IAM policy

Reviewer 2: Can we include an example policy document with this permission as well? Like you did above for the discovery service policy.

in order to receive commands from the Discovery Node. More information
for IAM best practices on EC2 instances managed by Systems Manager can be found [here](https://aws.amazon.com/blogs/mt/applying-managed-instance-policy-best-practices/)

> ## Step 4/4. Configure Teleport to discover EC2 instances

Reviewer 2: This step needs to also explain how to generate a join token for the discovery service, cause it needs to include "Discovery" role in it, correct? Like we do for other services e.g. https://goteleport.com/docs/database-access/guides/mongodb-self-hosted/#set-up-the-teleport-database-service.

In order to enable EC2 instance discovery the `discovery_service` must
include an AWS section configured with at least one entry:

```yaml

> discovery_service:

Reviewer 2: I would give the full yaml config example so users can copy-paste.

  enabled: "yes"
  aws:
   - types: ["ec2"]
     regions: ["eu-central-1"]
     tags:
       "env": "prod" # Match EC2 instances where tag:env=prod
```

Full documentation on EC2 discovery configuration can be found [here](../../reference/config.mdx)

> Once Teleport is configured for discovery, it can be started and EC2

Reviewer 2: I think we're missing a step that explains how to override the default install script. Maybe add an optional step "5" that explains what the default install script does and how to replace it if someone wants.

---
title: Configure Teleport to Automatically Enroll EC2 instances
description: How to configure Teleport to automatically enroll EC2 instances.
---

Teleport can be configured to run a discovery service that will
connect to EC2 and automatically discover and enroll EC2 instances
matching configured labels. It will then execute an install script on
these discovered instances using AWS Systems Manager, that will install
Teleport, start it and join the cluster.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- AWS account with EC2 instances and permissions to create and attach IAM policies as well as permission to create Systems Manager documents.
- EC2 instances running Ubuntu/Debian/RHEL/Amazon Linux 2 if making use of the default Teleport install script.
- A running Teleport Cluster.

## Step 1/4. Create an EC2 invite token

> When discovering EC2 nodes Teleport makes use of IAM invite tokens for

Reviewer 2: ```suggestion
When discovering EC2 nodes, Teleport makes use of IAM invite tokens for
```

authenticating joining Nodes.

Create the file `token.yaml`:
```yaml
# token.yaml
---
kind: token
version: v2
metadata:
  # the token name is not a secret because instances must prove that they are
  # running in your AWS account to use this token
  name: aws-discovery-iam-token
  # set a long expiry time, the default for tokens is only 30 minutes
  expires: "3000-01-01T00:00:00Z"
spec:
  # use the minimal set of roles required
  roles: [Node]

  # set the join method allowed for this token
  join_method: iam

  allow:
  # specify the AWS account which nodes may join from
  - aws_account: "123456789"
```

Add the token to the Teleport cluster with:
```
tctl create -f token.yaml
```

More documentation on IAM invite tokens can be found [here](../../management/guides/joining-nodes-aws-iam.mdx)

## Step 2/4. Create an AWS Systems Manager Document

Teleport makes use of AWS Systems Manager documents to execute commands necessary to
install and start Teleport on discovered EC2 instances.

> A document with the following content should be created, named `TeleportDiscoveryInstaller`:

Reviewer 2: Mention that it can be named differently but you'd need to update your auto-discovery configuration to use a different name.

```yaml
---
schemaVersion: '2.2'
description: aws:runShellScript
parameters:
  token:
    type: String
    description: "(Required) The Teleport invite token to use when joining the cluster."
  scriptName:
    type: String
    description: "(Required) The Teleport instalelr script to use when joining the cluster."
mainSteps:
- action: aws:downloadContent
  name: downloadContent
  inputs:
    sourceType: "HTTP"
    destinationPath: "/tmp/installTeleport.sh"
    sourceInfo:
      url: "https://teleport.example.com/webapi/scripts/{{ scriptName }}"
- action: aws:runShellScript
  name: runShellScript
  inputs:
    timeoutSeconds: '300'
    runCommand:
      - /bin/sh /tmp/installTeleport.sh "{{ token }}"
```

Here, `teleport.example.com` should be substituted with the public proxy URL of the
Teleport cluster.

When Teleport executes the Systems Manager document it will provide the `token`
and `scriptName` parameters with defaults.

- `{{ scriptName }}` will by default evaluate to `default-installer`

> - `{{ token }}` will by default evauluate to `aws-discovery-iam-token`

Reviewer 2: ```suggestion
- `{{ token }}` will by default evaluate to `aws-discovery-iam-token`
```

                "ssm:UpdateInstanceAssociationStatus",
                "ssm:UpdateInstanceInformation"
            ],
            "Resource": "*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel"
            ],
            "Resource": "*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "ec2messages:AcknowledgeMessage",
                "ec2messages:DeleteMessage",
                "ec2messages:FailMessage",
                "ec2messages:GetEndpoint",
                "ec2messages:GetMessages",
                "ec2messages:SendReply"
            ],
            "Resource": "*"
        }
    ]
}
```

## Step 4/4. Configure Teleport to discover EC2 instances

The Discovery Service requires a valid auth token to connect to the cluster. Generate one by running the following command against your Teleport Auth Service and save it in `/tmp/token` on the node that will run the Discovery Service:

> ```
> tctl tokens add --type=discovery
> ```

Reviewer 2: ```suggestion
```code
$ tctl tokens add --type=discovery
```
```

In order to enable EC2 instance discovery the `discovery_service` must
include an AWS section configured with at least one entry:

```yaml
version: v2
teleport:
  join_params:
    token_name: "/tmp/token"
    method: token
  auth_servers:
  - "teleport.example.com:3080"
auth_service:
  enabled: off
proxy_service:
  enabled: off
ssh_service:
  enabled: off
discovery_service:
  enabled: "yes"
  aws:
   - types: ["ec2"]
     regions: ["eu-central-1"]
     tags:
       "env": "prod" # Match EC2 instances where tag:env=prod
```

> Full documentation on EC2 discovery configuration can be found [here](../../reference/config.mdx)

Reviewer 2: ```suggestion
Full documentation on EC2 discovery configuration can be found [here](../../reference/config.mdx).
```

Teleport, start it and join the cluster.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- AWS account with EC2 instances and permissions to create and attach IAM policies as well as permission to create Systems Manager documents.
- EC2 instances running Ubuntu/Debian/RHEL/Amazon Linux 2 if making use of the default Teleport install script.
- A running Teleport Cluster.

## Step 1/5. Create an EC2 invite token

When discovering EC2 nodes, Teleport makes use of IAM invite tokens for
authenticating joining Nodes.

Create the file `token.yaml`:
```yaml
# token.yaml
---
kind: token
version: v2
metadata:
  # the token name is not a secret because instances must prove that they are
  # running in your AWS account to use this token
  name: aws-discovery-iam-token
  # set a long expiry time, the default for tokens is only 30 minutes
  expires: "3000-01-01T00:00:00Z"
spec:
  # use the minimal set of roles required
  roles: [Node]

  # set the join method allowed for this token
  join_method: iam

  allow:
  # specify the AWS account which nodes may join from
  - aws_account: "123456789"
```

Add the token to the Teleport cluster with:
```
tctl create -f token.yaml
```

> More documentation on IAM invite tokens can be found [here](../../management/guides/joining-nodes-aws-iam.mdx)

Reviewer 3: We should also move this link to a "Next steps" section, since it's not required for following this guide. We want to get users to the end of the guide as quickly as possible.

- `{{ scriptName }}` will by default evaluate to `default-installer`
- `{{ token }}` will by default evaluate to `aws-discovery-iam-token`

The document `TeleportDiscoveryInstaller` may be named differently
however the `discovery_service` section must be modified to include
the `ssm` section. For example:

```yaml
discovery_service:
  aws:
    - types: ["ec2"] 
      ssm:
        document_name: "OtherTeleportDiscoveryDocument"
```

## Step 3/5. Set up AWS IAM credentials

In order for Teleport to send commands to EC2 instances, the node
performing discovery requires an IAM policy with permissions to
execute Systems Manager commands and retrieve EC2 instances:

```js
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "ssm:SendCommand",
                "ec2:DescribeInstances",
                "ssm:GetCommandInvocation"
            ],
            "Resource": "*"
        }
    ]
}
```

All EC2 instances that are to be added to the Teleport cluster by the
Discovery Node must include the `AmazonSSMManagedInstanceCore` IAM policy
in order to receive commands from the Discovery Node. More information

> for IAM best practices on EC2 instances managed by Systems Manager can be found [here](https://aws.amazon.com/blogs/mt/applying-managed-instance-policy-best-practices/)

Reviewer 3: We should also move this link to a "Next steps" section. Users won't need to be familiar with IAM best practices to complete this guide, but this could be something they can learn about after completing the guide.

                "ssm:UpdateInstanceInformation"
            ],
            "Resource": "*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel"
            ],
            "Resource": "*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "ec2messages:AcknowledgeMessage",
                "ec2messages:DeleteMessage",
                "ec2messages:FailMessage",
                "ec2messages:GetEndpoint",
                "ec2messages:GetMessages",
                "ec2messages:SendReply"
            ],
            "Resource": "*"
        }
    ]
}
```

## Step 4/5. Configure Teleport to discover EC2 instances

The Discovery Service requires a valid auth token to connect to the cluster. Generate one by running the following command against your Teleport Auth Service and save it in `/tmp/token` on the node that will run the Discovery Service:

```code
tctl tokens add --type=discovery
```

> In order to enable EC2 instance discovery the `discovery_service` must

Reviewer 3: ```suggestion
In order to enable EC2 instance discovery the `discovery_service.aws` section must
```

include an AWS section configured with at least one entry:

```yaml
version: v2
teleport:
  join_params:
    token_name: "/tmp/token"
    method: token
  auth_servers:
  - "teleport.example.com:3080"
auth_service:
  enabled: off
proxy_service:
  enabled: off
ssh_service:
  enabled: off
discovery_service:
  enabled: "yes"
  aws:
   - types: ["ec2"]
     regions: ["eu-central-1"]
     tags:
       "env": "prod" # Match EC2 instances where tag:env=prod
```

Full documentation on EC2 discovery configuration can be found [here](../../reference/config.mdx).

Once Teleport is configured for discovery, it can be started and EC2
instances matching the tags specified in the AWS section will begin to
be added to the Teleport cluster automatically.

## Step 5/5. [Optional] Customize the default installer script

> To customize the default installer script, execute the following

Reviewer 3: ```suggestion
To customize the default installer script, execute the following command:

```

```code

> tctl get installer/default-installer > teleport-default-installer.yaml

Reviewer 3: ```suggestion
$ tctl get installer/default-installer > teleport-default-installer.yaml
```

```
The resulting `teleport-default-installer.yaml` can the be edited to
change what gets executed when enrolling discovered EC2 instances.

After making the desired changes to the default installer, the
resource can be updated by executing:

```code

> tctl create teleport-default-installer.yaml

Reviewer 3: ```suggestion
$ tctl create teleport-default-installer.yaml
```

```

> Multiple installer resources can exist and be specified in the

Reviewer 3: ```suggestion
Multiple `installer` resources can exist and be specified in the
```

`install/script_name` section:
```yaml
discovery_service:
  aws:
    - types: ["ec2"] 
      tags:
       - "env": "prod"
      install: # optional section when default-installer is used.
        script_name: "default-installer" 
    - types: ["ec2"] 
      tags:
       - "env": "devel"
      install:
        script_name: "devel-installer"		
```

---

> The installer resource has the following templating options:

Reviewer 3: ```suggestion
The `installer` resource has the following templating options:
```

> The installer resource has the following templating options:

Reviewer 3: Also, I'm not sure I know what this means—can you include an example of adding these template variables to an `installer` resource?

- `{{ .MajorVersion }}`: the major version of Teleport to use when
  installing from the repository.

> - `{{ .PublicProxyAddr }}`: the public address of the Teleport Proxy to connect to.

Reviewer 3: ```suggestion
- `{{ .PublicProxyAddr }}`: the public address of the Teleport Proxy Service to connect to.
```

The default installer will take the following actions:
- Add an official Teleport repository to supported Linux distributions.

> - Install Teleport via apt or yum.

Reviewer 3: ```suggestion
- Install Teleport via `apt` or `yum`.
```

> - Generate the Teleport config file and write it to /etc/teleport.yaml.

Reviewer 3: ```suggestion
- Generate the Teleport config file and write it to `/etc/teleport.yaml`.
```

- Enable and start the Teleport service.

> The default installer can be found [here](todo(amk):wait-for-pr-with-change-to-merge).

Reviewer 3: As with my other comments, we should make "here" more descriptive.

---
title: Configure Teleport to Automatically Enroll EC2 instances
description: How to configure Teleport to automatically enroll EC2 instances.
---

Teleport can run as a Discovery Node by configuring the discovery
service. This will connect to EC2 and automatically discover and
enroll EC2 instances matching configured labels. It will then execute
an install script on these discovered instances using AWS Systems
Manager that will install Teleport, start it and join the cluster.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- AWS account with EC2 instances and permissions to create and attach IAM policies as well as permission to create Systems Manager documents.

> - EC2 instances running Ubuntu/Debian/RHEL/Amazon Linux 2 if making use of the default Teleport install script. 

Reviewer 3: ```suggestion
- EC2 instances running Ubuntu/Debian/RHEL/Amazon Linux 2 if making use of the default Teleport install script. (For other Linux distributions, you can install Teleport manually.)
```
Feel free to adjust this suggestion—I think we need explicit instructions for unlisted distributions so users with those distributions don't get confused here.

## Step 1/5. Create an EC2 invite token

When discovering EC2 instances, Teleport makes use of IAM invite tokens for
authenticating joining Nodes.

Create a file called `token.yaml`:
```yaml
# token.yaml
---
kind: token
version: v2
metadata:
  # the token name is not a secret because instances must prove that they are
  # running in your AWS account to use this token
  name: aws-discovery-iam-token
  # set a long expiry time, as the default for tokens is only 30 minutes
  expires: "3000-01-01T00:00:00Z"
spec:
  # use the minimal set of roles required
  roles: [Node]

  # set the join method allowed for this token
  join_method: iam

  allow:

>   # specify the AWS account which nodes may join from

Reviewer 3: ```suggestion
  # specify the AWS account which Nodes may join from
```

     --document-format YAML
```

When Teleport executes the Systems Manager document it will provide the `token`
and `scriptName` parameters with defaults.

- `{{ scriptName }}` will by default evaluate to `default-installer`
- `{{ token }}` will by default evaluate to `aws-discovery-iam-token`

The document `TeleportDiscoveryInstaller` may be named differently
however the `discovery_service` section must be modified to include
the `ssm` section. For example:

```yaml
discovery_service:
  aws:
    - types: ["ec2"] 
      ssm:
        document_name: "OtherTeleportDiscoveryDocument"
```

## Step 3/5. Set up AWS IAM credentials

In order for Teleport to send commands to EC2 instances, the instance
performing discovery requires an IAM policy with permissions to
execute Systems Manager commands and retrieve EC2 instances:

```js
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "ssm:SendCommand",
                "ec2:DescribeInstances",
                "ssm:GetCommandInvocation"
            ],
            "Resource": "*"
        }
    ]
}
```

All EC2 instances that are to be added to the Teleport cluster by the

> Discovery Node must include the `AmazonSSMManagedInstanceCore` IAM policy

Reviewer 3: I would recommend `s/Discovery Node/Discovery Service/` to align with the way we name other Teleport services.

                "ssm:ListAssociations",
                "ssm:ListInstanceAssociations",
                "ssm:PutInventory",
                "ssm:PutComplianceItems",
                "ssm:PutConfigurePackageResult",
                "ssm:UpdateAssociationStatus",
                "ssm:UpdateInstanceAssociationStatus",
                "ssm:UpdateInstanceInformation"
            ],
            "Resource": "*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel"
            ],
            "Resource": "*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "ec2messages:AcknowledgeMessage",
                "ec2messages:DeleteMessage",
                "ec2messages:FailMessage",
                "ec2messages:GetEndpoint",
                "ec2messages:GetMessages",
                "ec2messages:SendReply"
            ],
            "Resource": "*"
        }
    ]
}
```

## Step 4/5. Configure Teleport to discover EC2 instances

> The discovery service requires a valid auth token to connect to the cluster. Generate one by running the following command against your Teleport Auth Service and save it in `/tmp/token` on the node that will run the discovery service:

Reviewer 3: ```suggestion
The Teleport Discovery Service requires a valid auth token to connect to the cluster. Generate one by running the following command against your Teleport Auth Service and save it in `/tmp/token` on the Node that will run the Discovery Service:
```

```code
$ tctl tokens add --type=discovery
```

In order to enable EC2 instance discovery the `discovery_service.aws` section must
include at least one entry:

```yaml
version: v2
teleport:
  join_params:
    token_name: "/tmp/token"
    method: token
  auth_servers:
  - "teleport.example.com:3080"
auth_service:
  enabled: off
proxy_service:
  enabled: off
ssh_service:
  enabled: off
discovery_service:
  enabled: "yes"
  aws:
   - types: ["ec2"]
     regions: ["eu-central-1"]
     tags:
       "env": "prod" # Match EC2 instances where tag:env=prod
```

Once Teleport is configured for discovery, it can be started and EC2
instances matching the tags specified in the AWS section will begin to
be added to the Teleport cluster automatically.

## Step 5/5. [Optional] Customize the default installer script

> To customize the default installer script, execute the following command:

Reviewer 3: ```suggestion
To customize the default installer script, execute the following command on your workstation:
```
We mention a few different hosts in this page, so we should indicate which one you should use to run a given command.
